docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,874 @@
+
+```ts
+interface HotContext {
+  file: string
+  timestamp: number
+  modules: Array<ModuleNode>
+  read: () => string | Promise<string>
+  server: ViteDevServer
+}
+```
+
+- `this.environment` is the module execution environment where a file update is currently being processed.
+
+- `modules` is an array of modules in this environment that are affected by the changed file. It's an array because a single file may map to multiple served modules (e.g. Vue SFCs).
+
+- `read` is an async read function that returns the content of the file. This is provided because, on some systems, the file change callback may fire too fast before the editor finishes updating the file, and direct `fs.readFile` will return empty content. The read function passed in normalizes this behavior.
+
+The hook can choose to:
+
+- Filter and narrow down the affected module list so that the HMR is more accurate.
+
+- Return an empty array and perform a full reload:
+
+  ```js
+  hotUpdate({ modules, timestamp }) {
+    if (this.environment.name !== 'client')
+      return
+
+    this.environment.hot.send({ type: 'full-reload' })
+    // Invalidate modules manually
+    const invalidatedModules = new Set()
+    for (const mod of modules) {
+      this.environment.moduleGraph.invalidateModule(
+        mod,
+        invalidatedModules,
+        timestamp,
+        true
+      )
+    }
+    return []
+  }
+  ```
+
+- Return an empty array and perform complete custom HMR handling by sending custom events to the client:
+
+  ```js
+  hotUpdate() {
+    if (this.environment.name !== 'client')
+      return
+
+    this.environment.hot.send({
+      type: 'custom',
+      event: 'special-update',
+      data: {}
+    })
+    return []
+  }
+  ```
+
+  Client code should register the corresponding handler using the [HMR API](./api-hmr) (this could be injected by the same plugin's `transform` hook):
+
+  ```js
+  if (import.meta.hot) {
+    import.meta.hot.on('special-update', (data) => {
+      // perform custom update
+    })
+  }
+  ```
+
+### Per-environment Plugins
+
+There is a new `create` hook that plugins can define to lazily create per-environment plugins.
+
+```js
+function perEnvironmentPlugin() {
+  return {
+    name: 'per-environment-plugin',
+    // Return a plugin, an array, a Promise, or a falsy value for each environment
+    create(environment: Environment) {
+      if (!passesCondition(environment)) {
+        return undefined
+      }
+      return [
+        createEnvironmentPlugin(environment),
+        otherPlugin(environment)
+      ]
+    }
+  }
+}
+```
+
+## `ModuleRunner`
+
+A module runner is instantiated in the target runtime. All APIs in the next section are imported from `vite/module-runner` unless stated otherwise. This export entry point is kept as lightweight as possible, only exporting the minimal needed to create runners in the
```

Comment:
NIT: I think this sentence is either not complete or malformed.

```suggestion
A module runner is instantiated in the target runtime. All APIs in the next section are imported from `vite/module-runner` unless stated otherwise. This export entry point is kept as lightweight as possible, only exporting the minimal needed to create module runners.
```
or
```suggestion
A module runner is instantiated in the target runtime. All APIs in the next section are imported from `vite/module-runner` unless stated otherwise. This export entry point is kept as lightweight as possible, only exporting the minimal needed to create runners in the target runtime.
```

Reply:
Thanks!